code/iaas/logic/src/main/java/io/cattle/platform/process/host/HostActivate.java

```diff
@@ -1,19 +1,77 @@
 package io.cattle.platform.process.host;
 
+import io.cattle.iaas.cluster.process.lock.ClusterLock;
+import io.cattle.iaas.cluster.service.ClusterManager;
+import io.cattle.platform.core.constants.ClusterConstants;
+import io.cattle.platform.core.constants.InstanceConstants;
 import io.cattle.platform.core.model.Agent;
 import io.cattle.platform.core.model.Host;
+import io.cattle.platform.core.model.Instance;
 import io.cattle.platform.engine.handler.HandlerResult;
 import io.cattle.platform.engine.process.ProcessInstance;
 import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.lock.LockCallbackNoReturn;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.resource.ResourceMonitor;
+import io.cattle.platform.object.resource.ResourcePredicate;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.object.util.DataUtils;
 import io.cattle.platform.process.base.AbstractDefaultProcessHandler;
 
+import javax.inject.Inject;
 import javax.inject.Named;
 
+import org.apache.commons.lang3.StringUtils;
+
 @Named
 public class HostActivate extends AbstractDefaultProcessHandler {
 
+    @Inject
+    ResourceMonitor resourceMonitor;
+
+    @Inject
+    ClusterManager clusterManager;
+
+    @Inject
+    LockManager lockManager;
+
     @Override
     public HandlerResult handle(final ProcessState state, ProcessInstance process) {
+        if (objectManager.isKind(state.getResource(), ClusterConstants.KIND)) {
+            return handleCluster(state, process);
+        } else {
+            return handleHost(state, process);
+        }
+    }
+
+    private HandlerResult handleCluster(final ProcessState state, ProcessInstance process) {
+        final Host cluster = (Host)state.getResource();
+        lockManager.lock(new ClusterLock(cluster), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                // short term optimization to avoid updating cluster object unnecessarily
+                // since we're just currently only supporting file:// discoverySpec
+                if (StringUtils.isEmpty(DataAccessor.fieldString(cluster, ClusterConstants.DISCOVERY_SPEC))) {
+                    DataUtils.getWritableFields(cluster).put(ClusterConstants.DISCOVERY_SPEC, "file:///etc/cluster-hosts.conf");
+                    objectManager.persist(cluster);
+                }
+
+                Instance clusterServerInstance = clusterManager.createClusterServerInstance(cluster);
```

Comment:
This should be moved to the ClusterManagerImpl as creating the cluster service instance is a prereq for server config update